### PR TITLE
feat: introduce package UIDs for improved vulnerability mapping

### PR DESCRIPTION
**Pull request by knqyf263**
## Description
Currently, Trivy outputs [Packages](https://github.com/aquasecurity/trivy/blob/a5d485cf8a36f3f50b9fe4c3b58641eb364b8ed8/pkg/types/report.go#L108) and [Vulnerabilities](https://github.com/aquasecurity/trivy/blob/a5d485cf8a36f3f50b9fe4c3b58641eb364b8ed8/pkg/types/report.go#L109) separately in JSON format, with detected vulnerabilities including a subset of package metadata such as name and version. However, identifying the original package from this limited metadata can be challenging in certain scenarios:

- When multiple lock files use the exact same package name and version (but have different dependencies or line numbers)
- When the same package is installed in multiple paths

To accurately associate detected vulnerabilities with their corresponding packages in these cases, this PR introduces the concept of package UIDs. These UIDs are calculated from [the Package struct](https://github.com/aquasecurity/trivy/blob/a5d485cf8a36f3f50b9fe4c3b58641eb364b8ed8/pkg/fanal/types/artifact.go#L79-L118) and serve as unique identifiers for each package. It's important to note that these values (hashes) are not the same as digests used for ensuring integrity; they are solely intended for uniquely identifying packages within Trivy.
One advantage of using hashes instead of UUIDs is that the same hash value will be generated across multiple runs for the same package.

## Checklist
- [x] I've read the [guidelines for contributing](https://aquasecurity.github.io/trivy/latest/community/contribute/pr/) to this repository.
- [x] I've followed the [conventions](https://aquasecurity.github.io/trivy/latest/community/contribute/pr/#title) in the PR title.
- [x] I've added tests that prove my fix is effective or that my feature works.
- [ ] I've updated the [documentation](https://github.com/aquasecurity/trivy/blob/main/docs) with the relevant information (if needed).
- [ ] I've added usage information (if the PR introduces new options)
- [ ] I've included a "before" and "after" example to the description (if the PR is a user interface change).

### CODE REVIEW

**Comment by knqyf263** [comment]
@DmitriyLewen Please let me know if you have a better idea.

File: pkg/iac/providers/aws/iam/actions.go
Comment (knqyf263) [inline pkg/iac/providers/aws/iam/actions.go:1]: Seems like IDE automatically formatted it. I'll revert it.

File: pkg/fanal/test/integration/testdata/goldens/packages/alpine-310.json.golden
Comment (knqyf263) [inline pkg/fanal/test/integration/testdata/goldens/packages/alpine-310.json.golden:1]: I have no idea what happened... I have reverted `pkg/iac/providers/aws/iam/actions.go`, checked the difference and committed it.

File: pkg/iac/providers/aws/iam/actions.go
Comment (knqyf263) [reply pkg/iac/providers/aws/iam/actions.go:1]: reverted https://github.com/aquasecurity/trivy/pull/6583/commits/a188c94f176703b8cda475eed2fcbc1fc510d6db

**Comment by knqyf263** [comment]
@DmitriyLewen I didn't come up with a good idea of where we should calculate the UID. Please let me know if you have a better one.

**Comment by DmitriyLewen** [comment]
I think you chose a good place. we create identifiers (purl, uid) in one place.

**Comment by knqyf263** [comment]
@DmitriyLewen It's finally ready for review!

**Review by DmitriyLewen — APPROVED** [review]
LGTM

**Pull request merged**